src/nvim/lua/xdiff.c

```diff
@@ -61,20 +66,90 @@ static int hunk_locations_cb(long start_a, long count_a, long start_b, long coun
   if (count_b > 0) {
     start_b += 1;
   }
-
   lua_State *lstate = (lua_State *)cb_data;
-  lua_createtable(lstate, 0, 0);
+  if (linematch) {
+    const char **diff_begin = xmalloc(sizeof(char *) * 2);
+    int *diff_length = xmalloc(sizeof(int) * 2);
+    diff_begin[0] = sma.ptr;
+    diff_length[0] = (int)count_a;
+    diff_begin[1] = smb.ptr;
+    diff_length[1] = (int)count_b;
+    // get the pointer to char of the start of the diff to pass it to linematch algorithm
+    for (int i = 0; i < 2; i++) {
+      for (int j = 0; j < (i ? start_b : start_a) - 1; j++) {
+        while (*diff_begin[i] != '\n') {
+          diff_begin[i]++;
+        }
+        diff_begin[i]++;
+      }
+    }
+    int decisions_length = 0;
+    int *decisions = NULL;
+    linematch_nbuffers(diff_begin, diff_length, 2,
+        &decisions_length, &decisions);
+
+    long lnuma = start_a, lnumb = start_b;
+
+    long hunkstarta = lnuma;
+    long hunkstartb = lnumb;
+    long hunkcounta = 0;
+    long hunkcountb = 0;
+    for (int i = 0; i < decisions_length; i++) {
+      if (i && (decisions[i - 1] != decisions[i])) {
+        // push the hunk to lua state
+        lua_createtable(lstate, 0, 0);
+        lua_pushinteger(lstate, hunkstarta);
+        lua_rawseti(lstate, -2, 1);
+        lua_pushinteger(lstate, hunkcounta);
+        lua_rawseti(lstate, -2, 2);
+        lua_pushinteger(lstate, hunkstartb);
+        lua_rawseti(lstate, -2, 3);
+        lua_pushinteger(lstate, hunkcountb);
+        lua_rawseti(lstate, -2, 4);
+        lua_rawseti(lstate, -2, (signed)lua_objlen(lstate, -2) + 1);
+
+        hunkstarta = lnuma;
+        hunkstartb = lnumb;
+        hunkcounta = 0;
+        hunkcountb = 0;
+        // create a new hunk
+      }
+      if (decisions[i] & (1 << 0)) {
+        lnuma++;
+        hunkcounta++;
+      }
+      if (decisions[i] & (1 << 1)) {
```

Comment:
```suggestion
      if (decisions[i] & 2) {
```
Then maybe define 2 to some meaningful name (`BMASK`)?

Reply:
I added a #define to call these COMPARED_BUFFER0 and COMPARED_BUFFER1